src/repo.rs

```diff
@@ -753,42 +795,89 @@ impl DocumentInfo {
     }
 
     /// Apply incoming sync messages,
-    /// returns whether the document changed due to applying the message.
-    fn receive_sync_message(&mut self, per_remote: HashMap<RepoId, VecDeque<SyncMessage>>) -> bool {
+    ///
+    /// # Returns
+    ///
+    /// A tuple of `(has_changes, commands)` where `has_changes` is true if the document changed as
+    /// a result of applying the sync message and `commands` is a list of changes requested by the
+    /// peer connections for this document (e.g. requesting authorization from the share policy).
+    fn receive_sync_message(
+        &mut self,
+        per_remote: HashMap<RepoId, VecDeque<SyncMessage>>,
+    ) -> (bool, Vec<PeerConnCommand>) {
+        let mut commands = Vec::new();
         let (start_heads, new_heads) = {
             let mut document = self.document.write();
             let start_heads = document.automerge.get_heads();
             for (repo_id, messages) in per_remote {
-                let sync_state = self.sync_states.entry(repo_id).or_default();
-
-                // TODO: remove remote if there is an error.
+                let conn = match self.peer_connections.entry(repo_id.clone()) {
+                    Entry::Vacant(entry) => {
+                        // if this is a new peer, request authorization
+                        commands.push(PeerConnCommand::RequestAuth(repo_id.clone()));
+                        entry.insert(PeerConnection::pending())
+                    }
+                    Entry::Occupied(entry) => entry.into_mut(),
+                };
                 for message in messages {
-                    document
-                        .automerge
-                        .receive_sync_message(sync_state, message)
-                        .expect("Failed to apply sync message.");
+                    conn.receive_sync_message(&mut document.automerge, message)
+                        .expect("Failed to receive sync message.");
                 }
             }
             let new_heads = document.automerge.get_heads();
             (start_heads, new_heads)
         };
-        start_heads != new_heads
+        (start_heads != new_heads, commands)
+    }
+
+    /// Promote a peer awaiting authorization to a full peer
+    ///
+    /// Returns any messages which the peer sent while we were waiting for authorization
+    fn promote_pending_peer(&mut self, repo_id: &RepoId) -> Option<Vec<SyncMessage>> {
+        if let Some(PeerConnection::PendingAuth { received_messages }) =
+            self.peer_connections.remove(repo_id)
+        {
+            self.peer_connections
+                .insert(repo_id.clone(), PeerConnection::Accepted(SyncState::new()));
```

Comment:
Why not apply all received messages to the sync state here?

Reply:
Ah I see, because it needs to go through `sync_documents` to generate potential outgoing sync messages.